eth2util/deposit/deposit_data.go

```diff
@@ -18,142 +18,219 @@ package deposit
-	ssz "github.com/ferranbt/fastssz"
+	"github.com/ethereum/go-ethereum/common"
 
 	"github.com/obolnetwork/charon/app/errors"
+	"github.com/obolnetwork/charon/app/z"
 )
 
-const depositAmt = 32000000000
+var (
+	// https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/validator.md#eth1_address_withdrawal_prefix
+	eth1AddressWithdrawalPrefix = byte(0x01)
 
-// depositData contains all the information required for activating validators on the Ethereum Network.
-// Ref: https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#depositdata
-type depositData struct {
-	// PubKey is the group public key for a Distributed Validator.
-	PubKey eth2p0.BLSPubKey
+	// the amount of ether in gwei required to activate a validator.
+	validatorAmt uint64 = 32000000000
 
-	// Amount is the amount of Eth needed to activate a validator.
-	Amount eth2p0.Gwei
+	// zeroes11 refers to a zeroed out 11 byte array.
+	zeroBytes11 = []byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
 
-	// Eth1WithdrawalAddress is the Ethereum withdrawal address.
-	Eth1WithdrawalAddress string
+	// zeroes refers to a zeroed out 32 byte array.
+	zeroBytes32 = []byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
 
-	// DepositMessageRoot is the hash tree root of DepositMessage.
-	DepositMessageRoot eth2p0.Root
+	// DOMAIN_DEPOSIT. See spec: https://benjaminion.xyz/eth2-annotated-spec/phase0/beacon-chain/#domain-types
+	depositDomainType = []byte{0x03, 0x00, 0x00, 0x00}
+)
+
+// GetMessageRoot returns both the hash root and the signing root of the deposit message.
+func GetMessageRoot(pubkey eth2p0.BLSPubKey, withdrawalCreds [32]byte, forkVersion eth2p0.Version) (eth2p0.Root, eth2p0.Root, error) {
+	dm := eth2p0.DepositMessage{
+		PublicKey:             pubkey,
+		WithdrawalCredentials: withdrawalCreds[:],
+		Amount:                eth2p0.Gwei(validatorAmt),
+	}
+
+	hashRoot, err := dm.HashTreeRoot()
+	if err != nil {
+		return eth2p0.Root{}, eth2p0.Root{}, errors.Wrap(err, "deposit message hash root")
+	}
 
-	// Signature is constructed from DepositMessageRoot combined with DOMAIN_DEPOSIT.
-	Signature eth2p0.BLSSignature
+	signingRoot, err := GetSigningRoot(forkVersion, hashRoot)
+	if err != nil {
+		return eth2p0.Root{}, eth2p0.Root{}, errors.Wrap(err, "deposit message root")
+	}
 
-	// ForkVersion identifies the network/chainID.
-	ForkVersion eth2p0.Version
+	return hashRoot, signingRoot, nil
 }
 
-func (d depositData) HashTreeRoot() ([32]byte, error) {
-	b, err := ssz.HashWithDefaultHasher(d)
+// GetDataRoot returns both the hash root and the signing root of the deposit data.
+func GetDataRoot(pubkey eth2p0.BLSPubKey, withdrawalCreds [32]byte, sig eth2p0.BLSSignature, forkVersion eth2p0.Version) (eth2p0.Root, eth2p0.Root, error) {
```

Comment:
only one of the two responses is ever used, so suggest only retuning that one thing

Reply:
yes corrected that